source/client-side-encryption/client-side-encryption.rst

```diff
@@ -242,70 +334,318 @@ MongoClient Changes
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The `extraOptions` parameter to AutoEncryptionOpts_ relate to `mongocryptd`
+and `csfle`, with more detail described in the Implementation_ section:
+
+.. code-block:: typescript
+
+   interface ExtraOptions {
+      mongocryptdURI: string;
+      mongocryptdBypassSpawn: boolean;
+      mongocryptdSpawnPath: string,
+      mongocryptdSpawnArgs: string[];
+      csfleSearchPaths: string[];
+      csflePathOverride: string | undefined;
+      csfleRequired: boolean;
+   }
+
+Drivers MUST implement extraOptions in a way that allows
+deprecating/removing options in the future without an API break, such as
+with a BSON document or map type instead of a struct type with fixed
+fields.
+
+.. seealso:: `Why are extraOptions and kmsProviders maps?`_.
+
+.. rubric:: Options
+
+.. option:: extraOptions.mongocryptdURI
+
+   :type: :ts:`string`
+   :default: :ts:`"mongodb://localhost:27020"`
+
+   Allows the user to set the URI to use when connecting to `mongocryptd`.
+
+.. option:: extraOptions.mongocryptdBypassSpawn
 
-keyVaultNamespace
-^^^^^^^^^^^^^^^^^
-The key vault collection namespace refers to a collection that contains all
-data keys used for encryption and decryption (aka the key vault collection).
-Data keys are stored as documents in a special MongoDB collection. Data
-keys are protected with encryption by a KMS provider (AWS KMS, Azure key
-vault, GCP KMS, a local master key, or KMIP).
+   :type: :ts:`boolean`
+   :default: |false|
 
-keyVaultClient
-^^^^^^^^^^^^^^
-The key vault collection is assumed to reside on the same MongoDB
-cluster as indicated by the connecting URI. But the optional
-keyVaultClient can be used to route data key queries to a separate
-MongoDB cluster.
+   Disables automatic spawning of `mongocryptd`. Refer:
+   `Managing mongocryptd`_
 
-If a ``keyVaultClient`` is not passed, and the parent ``MongoClient`` is
-configured with a limited ``maxPoolSize``, the ``keyVaultClient`` is set to an
-internal ``MongoClient``. See `keyVaultClient, metadataClient, and the internal
-MongoClient`_ for configuration behavior.
+.. option:: extraOptions.mongocryptdSpawnArgs
+
+   :type: :ts:`string[]`
+   :default: :ts:`["--idleShutdownTimeoutSecs=60"]`
+
+   Allow the user to specify additional command-line options for spawning
+   `mongocryptd` Refer: `Managing mongocryptd`_
+
+.. option:: extraOptions.mongocryptdSpawnPath
+
+   :type: :ts:`string`
+   :default: :ts:`""`
+
+   Used for spawning. Defaults to empty string and spawns `mongocryptd` from
+   system path. Refer: `Managing mongocryptd`_
+
+.. option:: extraOptions.csfleSearchPaths
+
+   :type: :ts:`string[]`
+   :default: :ts:`["$ORIGIN"]`
+
+   .. |opt-paths| replace:: :option:`extraOptions.csfleSearchPaths`
+
+   Allow the user to specify directories in which to search for `csfle`. Refer:
+
+   - :ref:`cse.csfle.search-paths`
+   - :ref:`cse.csfle.path-resolution`
+   - :ref:`cse.enabling-csfle`
+
+.. option:: extraOptions.csflePathOverride
```

Comment:
Can users use `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH`, and `PATH` to set a path to csfle? If "yes", can `csflePathOverride` be removed?

Drivers will not be packaging `csfle`. AFAIK the main use case of `csflePathOverride` is mongosh. @addaleax can mongosh use `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH`, and `PATH` to specify the path to `csfle`?

Reply:
> Can users use `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH`, and `PATH` to set a path to csfle? If "yes", can `csflePathOverride` be removed?

I assume that, technically, the answer to the first question is “yes”. However, users may be reluctant – I certainly would be – to modify global state that affects external code loading (like these environment variables) for what is potentially a small part of their application.

If users were using these environment variables, I think they would replace `csfleSearchPaths` rather than `csflePathOverride`?

> AFAIK the main use case of `csflePathOverride` is mongosh. 

`csflePathOverride` is going to be used by mongosh, but I don’t see any reason why an application that bundles the shared library in a specific location and consequently can know the exact path to it wouldn’t want to use it.

> can mongosh use `LD_LIBRARY_PATH`, `DYLD_LIBRARY_PATH`, and `PATH` to specify the path to `csfle`?

No, the filename would still mismatch.

Reply:
> I assume that, technically, the answer to the first question is “yes”. However, users may be reluctant – I certainly would be – to modify global state that affects external code loading (like these environment variables) for what is potentially a small part of their application.

That is a good point. It may not be a desirable option for users.

> No, the filename would still mismatch.

Good point. `csflePathOverride` is needed to specify the filename.

--

To simplify changes for drivers, is it an option is to only require the Node.JS driver support `csflePathOverride`? And not require drivers support `csfleSearchPaths`.

Reply:
What is the use case for an application (presumably mongosh or Compass?) needing a different file name than the default one?  If it's a narrow enough use case that we don't think likely to be common in normal applications, then we could leave it out of the spec and just implement it in the Node driver.

Reply:
I'm not sure how reluctant users will be to set `LD_LIBRARY_PATH`.  Any search path that is specified via this option would be externalized in _some_ way: you wouldn't want to embed the path(s) directly in the application.  If that's the case, LD_LIBRARY_PATH, etc. seems as good a mechanism as any for that, and it has the advantage of being a standard one that is already supported by dlopen.

I think we could leave this option out and wait to see if it's requested by users.

Reply:
> To simplify changes for drivers, is it an option is to only require the Node.JS driver support `csflePathOverride`? 

And let other drivers use something like `LD_LIBRARY_PATH`? I don’t know. I’m personally mostly concerned about the Node.js driver of course, but we already did the work for supporting `csflePathOverride` in libmongocrypt, and that is imo the cleanest way to specify which library to load (because, as mentioned, it’s a local option controlling local behavior). Adding driver bindings for this is trivial – for the Node.js driver it was 7 SLOC. I’d definitely be in favor of doing it.

> And not require drivers support `csfleSearchPaths`.

I think this would be fine either way – it’s also not hard to add support for it in the way that this PR currently suggests, since the bulk of the work here is already done. On the other hand, I think it’s likely that the applications which would use `csfleSearchPaths` would be able to use `csflePathOverride` just as well.

> What is the use case for an application (presumably mongosh or Compass?) needing a different file name than the default one? 

In mongosh we want to avoid possible conflicts of the csfle shared library shipped with mongosh and a possible system installation, and avoid accidentally making csfle available to users who have not specifically opted into using it by downloading it separately from mongosh (and thus entering into the enterprise license agreement). It’s the same thing we currently do with mongocryptd.

> you wouldn't want to embed the path(s) directly in the application

Why not?

> and it has the advantage of being a standard one that is already supported by dlopen.

This still feels like we’re asking for trouble. `LD_LIBRARY_PATH` needs to be set before the process starts, and its functionality can be disabled entirely depending on your selinux config. Users will set `DYLD_LIBRARY_PATH` without knowing that what they probably wanted to set was `DYLD_FALLBACK_FRAMEWORK_PATH`. We might get bug reports about `PATH` not working on Windows and have to tediously figure out that yet another, completely unrelated, part of their application overrode `PATH` after it had already been set, or accidentally used `:` instead of `;`.

I know this is all hypothetical, but I really think we should tell users to go with `csfleOverridePath`. They can still use the OS-specific library loading env vars if they really want to, but an explicit option should give us the best behavior in terms of consistency and diagnostics and is, for better or for worse, already implemented and easy to add driver bindings for.

Reply:
Either way, setting one of these environment variables only has an effect on the behavior of the system's library search mechanism, and does not force the behavior i.e. it tells the loader to look in the given directories before it's own defaults, whereas `csfleOverridePath` would be an exact filepath given to `dlopen`/`LoadLibrary` and will be a clear hard-error if it names a non-existent path. If users specify any such environment variable, that will effectively inject those paths when the `$SYSTEM` search is performed.

Reply:
I've pushed changes that remove `csfleSearchPaths` and I've opened a PR on `libmongocrypt` that removes the functionality as well. This necessitates a new option to disable `csfle` since "don't specify any search paths" is no longer an option.

Reply:
Are we certain it's a requirement to be able to disable it explicitly?

Reply:
I seem to recall someone addressing that aspect explicitly, but I cannot find where it was. We can remove the option as well.

Reply:
I think the prior discussion you are referring to was https://github.com/mongodb/specifications/pull/1165#discussion_r838338960?

I think it makes sense to give the users the option to explicitly specify not using the shared library in some way in order to get consistent behavior across installations. It wouldn’t need to be an extra option in every driver – e.g. in Node.js it could be a tristate `'require-csfle' | 'try-load-csfle' | 'ignore-csfle'` or similar.

Reply:
> Are we certain it's a requirement to be able to disable it explicitly?

It is not a requirement to have an option for disabling searching for `csfle`.

It may be a good precaution. If `csfle` and `mongocryptd` are both on the system, `csfleDisabled` may be the only way to force using `mongocryptd`. Maybe it could be useful in the event that `csfle` has a regression not present on `mongocryptd`. Are there other use-cases?

Reply:
I am OK with the currently proposed options (`csflePath`, `csfleRequired`, and `csfleDisabled`). I think arguably we could remove `csfleRequired` and `csfleDisabled`, but they may provide value to users in the event that mongocryptd and csfle are both on the system.